Add initial IPv6 support for Cloudflare

robertcampion opened the pull request:
This PR adds support for querying the current IPv6 address and updating AAAA records for most providers.

Fully supported:

- [x] Cloudflare
- [x] Dynu
- [x] GoDaddy
- [x] Hurricane Electric
- [x] No-IP
- [x] Porkbun

Not supported:

- [ ] Namecheap: their DDNS API [doesn't support IPv6](https://www.namecheap.com/support/knowledgebase/article.aspx/29/11/how-to-dynamically-update-the-hosts-ip-with-an-https-request/#:~:text=Currently%2C%20IPv6%20is%20not%20supported%20by%20our%20Dynamic%20DNS). Their full API does, but it requires [whitelisting your IP address](https://www.namecheap.com/support/knowledgebase/article.aspx/9739/63/api-faq/#:~:text=Once%20API%20access%20is%20enabled%2C%20you%20should%20whitelist%20at%20least%20one%20IP%20before%20your%20API%20access%20begin%20to%20work%2E%20Please%20keep%20in%20mind%20that%20only%20IPv4%20addresses%20can%20be%20used%2E) to access it, which means it is unusable for DDNS.

For the config file, I adopted the suggestion from #386 of adding an `ip_types` field that can be set to a list containing `"4"` or `"6"` (or both), for example:

```toml
[[domains]]
type = "cloudflare"
token = "..."
zone = "example.com"
records = ["subdomain.example.com"]
ip_types  = ["4", "6"]
```

`ip_types` will default to `["4"]`, so that current configurations have the same behavior.

`DomainConfig` gained a method `get_ip_types(&self) -> Vec<IpType>` to provide a uniform interface for the different configuration types. For configs with IPv6 support it clones `ip_types` and the other configs return `[IpType::V4]` so their behavior is unchanged.

The other interfaces that had to change are the resolvers, which now take an `ip_type: IpType` argument and return `IpAddr` instead of `Ipv4Addr`. Similarly `update_provider` and `update_domains` now take `IpAddr` as their argument. Providers that only support IPv4 will return an error if an IPv6 address is provided.

The new data flow in `main` is:
* Call `get_ip_types` for all domain configs to get a list of what IP address types we need
* Get IP addresses of the identified types (or IPv4 only if no domain configs were specified)
* For each domain config, check which types were requested by this config and call `update_provider` on each of the corresponding addresses.

This does a little more work than needed for dual-stack configs (e.g. it will request the zone id once for each address type instead of once per config), but I think it's good for a first implementation. Theoretically since the code uses async we could process the IPv4 and IPv6 requests in parallel but that would involve more substantial changes to the code structure.

Resolves #386

robertcampion committed:
Add IP address type enum

* Also add ip_types field with new enum to CloudflareConfig

robertcampion committed:
Add tests for ip_types field

robertcampion committed:
Add IPv6 cloudflare support

* Use IpAddr instead of Ipv4Addr as the address type
* Add AAAA record handling to Cloudflare; panic on IPv6 for all other
  providers
* Add IPv6 handling to both resolvers
* Query required address types for each domain before getting IPs
* Call update_provider for each required address type (might duplicate
  work, e.g. fetching zone ID)

robertcampion committed:
resolve IP with empty config

robertcampion committed:
Contact resolver using same IP type

The resolver only knows the address that we contact it over, so to get
our IPv6 adress we need to contact it over IPv6

robertcampion committed:
handle returned IPv6 addresses

robertcampion committed:
Add tests for Cloudflare IPv6

robertcampion committed:
default to IPv4 only

robertcampion committed:
Continue when an address fails to resolve

In the case that we request both IPv4 and IPv6 addresses but are only
able to resolve one, we should continue with updating that address
intead of immediately exiting. Thanks to the fact we loop over the
successful addresses we don't need special error handling.

nickbabcock reviewed (commented):
Nice this is looking good, thanks for taking up this work.

robertcampion committed:
Don't panic on unsupported IP type

Although we currently only pass in IPs with supported types, we don't
want to crash the program if that happens in the future.

robertcampion committed:
Error on IP resolution failure

If no IP addresses could be resolved then we don't try and update any
DNS records. I didn't realize that with the original logic this means
the failure flag wasn't set and we exit with a success code.

robertcampion committed:
Disable live IPv6 tests by default

Although the GitHub runners have IPv6 support, they have no route to the
global internet, causing IPv6 address resolution to fail. We don't want
this to cause CI to fail, so we gate these tests behind the #[ignored]
attribute. The tests can still be run with

    cargo test -- --ignored

or

    cargo test -- --include-ignored

See actions/runner-images#668

nickbabcock commented:
> If/when I work on these, should I submit them as separate PRs or do you want them all grouped together?

I trust your judgement. If it makes sense to batch up the easy implementations, that sounds good to me.

robertcampion committed:
Change default address lookup to IPv4 only

Prevents CI failure on the IPv4-only GitHub Actions runners. Also keeps
the no-config behavior the same as before.

robertcampion committed:
Initial GoDaddy IPv6 support

robertcampion committed:
Dynu IPv6 support

robertcampion committed:
Hurricane Electric IPv6 support

robertcampion committed:
No-Ip IPv6 support

robertcampion committed:
Initial Porkbun IPv6 support

robertcampion committed:
Update documentation for ip_types

robertcampion committed:
Minor formatting

nickbabcock commented:
Thank you so much for this contribution. I've gone ahead and released this as part of 0.6.0
